lib/_stream_readable.js

```diff
@@ -1070,13 +1072,84 @@ Readable.prototype.wrap = function(stream) {
 };
 
 Readable.prototype[SymbolAsyncIterator] = function() {
-  if (createReadableStreamAsyncIterator === undefined) {
-    createReadableStreamAsyncIterator =
-      require('internal/streams/async_iterator');
+  let stream = this;
+
+  if (typeof stream.read !== 'function') {
+    // v1 stream
+    const src = stream;
+    stream = new Readable({
+      objectMode: true,
+      destroy(err, callback) {
+        destroyImpl.destroyer(src, err);
+        callback();
+      }
+    }).wrap(src);
   }
-  return createReadableStreamAsyncIterator(this);
+
+  const iter = createAsyncIterator(stream);
+  iter.stream = stream;
+  return iter;
 };
 
+async function* createAsyncIterator(stream) {
+  const state = stream._readableState;
+
+  if ((state && state.endEmitted) ||
+      (state && state.closeEmitted)) {
+    // TODO (ronag): What if errorEmitted?
+    // TODO (ronag): Wait for closeEmitted?
+    // TODO (ronag): if (closeEmitted && !endEmitted) ERR_PREMATURE_CLOSE?
+    stream.on('error', nop);
+    return;
+  }
+
+  let callback = nop;
+  let error;
+
+  stream
+    .on('readable', () => {
+      callback();
+      callback = nop;
+    })
+    .on('error', (err) => {
+      error = err;
+      callback();
+      callback = nop;
+    })
+    .on('end', () => {
+      error = error || null;
+      callback();
+      callback = nop;
+    })
+    .on('close', () => {
+      error = error || null;
+      // TODO (ronag): ERR_PREMATURE_CLOSE?
+      callback();
+      callback = nop;
+    });
+
+  try {
+    while (error === undefined) {
```

Comment:
I'm _probably_ being over cautious here but _technically_ someone can manually emit an `error` event (incorrectly) on the stream with `error === undefined` so it's _probably_ better to guard against that case or set the initial value for `error` to something the user can't set.

Reply:
that is a bit paranoid but fair enough :D. Changed.

Reply:
Should we throw undefined as well in this case?

Reply:
Probably :D?

Reply:
ugh, can't we just make EventEmitter throw if they try to emit a nully error? Or convert it here into `new Error()`.

Reply:
Probably not due to ecosystem breakage? We did this in bluebird at one point but had to revert.

Reply:
I'm unsure what to do here.

Reply:
Probably: Change the check to `=== someSymbolItIsOnlyEqualToInitially` and throw `undefined` if undefined is emitted as an error

Reply:
Fixed. But. I'm not sure we should fix this... we don't handle this in e.g. stream destroy...